core/bin/core_api.rs

```diff
@@ -1625,6 +1639,20 @@ async fn data_sources_documents_upsert(
         None => false,
     };
 
+    if payload.parents.get(0) != Some(&payload.document_id) {
+        info!(
+            data_source_id = data_source_id,
+            node_id = payload.document_id,
+            parents_0 = match payload.parents.get(0) {
+                None => "None",
+                Some(p) => p,
+            },
+            node_type = "document",
+            operation = "upsert",
+            "[KWSEARCH] invariant_first_parent_self"
+        );
+    }
+
```

Comment:
we should get rid of the one in `DataSource::upsert` no?

Reply:
did we add one?

Reply:
ah yes. removed.